### PR TITLE
Polkadot Analytics Platform (Stage 2)

### PR DESCRIPTION
**Pull request by rbrandao**
#### Project Abstract

The Polkadot Analytics Platform aims at building a comprehensive data analysis and visualization tool for the Polkadot ecosystem. The platform will allow users to retrieve and analyze data from various Polkadot-related sources (e.g., different parachains and components such as browser wallets), aligned with the POnto ontology [1, 2, 3]. Users will be able to specify their queries using a controlled natural language (CNL), and the platform will provide a query engine to process these queries. Additionally, the platform will provide a UI to support constructing queries and visualizing informative artifacts that represent query results. As well as support for composing customizable dashboards using these artifacts. 

In its current stage, the platform is composed of a knowledge base (KB) and its initial representation with the POnto ontology [4]. The current grant proposal focuses on populating this KB through the creation of semantic ETL pipelines, i.e. information extraction workflows that will extract, reuse and integrate data from different sources, aligning and structuring domain knowledge in the KB. We will create data extraction mechanisms to gather data from various Polkadot-related sources using Substrate interfaces (e.g. using Substrate-ETL), as well as offchain (e.g., using Github API).

[1] POnto source code: https://github.com/mobr-ai/POnto
[2] POnto documentation: https://www.mobr.ai/ponto
[3] POnto scientific paper:  https://github.com/mobr-ai/POnto/raw/main/deliverables/milestone3/article.pdf  
[4] Polkadot Analytics Platform source code: https://github.com/mobr-ai/PolkadotAnalytics 

This is a follow-up grant application for the project [A Knowledge-Oriented Approach to Enhance Integration and Communicability in the Polkadot Ecosystem](https://github.com/w3f/Grants-Program/pull/1420) and the project [A Polkadot Analytics Platform: Stage 1](https://github.com/w3f/Grants-Program/pull/1883) 

#### Grant [level](https://github.com/w3f/Grants-Program#level_slider-levels)

- [ ] **Level 1**:  Up to $10,000, 2 approvals
- [x] **Level 2**:  Up to $30,000, 3 approvals
- [ ] **Level 3**:  Unlimited, 5 approvals (for >$100k: Web3 Foundation Council approval)

#### Application Checklist

- [x] The [application template](https://github.com/w3f/Grants-Program/blob/master/applications/application-template.md) has been copied and aptly renamed (`project_name.md`).
- [x] I have read the [application guidelines](https://github.com/w3f/Grants-Program/blob/master/docs/Support%20Docs/grant_guidelines_per_category.md).
- [x] Payment details have been provided (bank details via email _or_ BTC, Ethereum (USDC/DAI) or Polkadot/Kusama (USDT) address in the application).
- [x] The software delivered for this grant will be released under an open-source license specified in the application.
- [x] The initial PR contains only one commit (squash and force-push if needed).
- [x] The grant will only be announced once the first milestone [has been accepted](https://github.com/w3f/Grant-Milestone-Delivery#process) (see the [announcement guidelines](https://github.com/w3f/Grants-Program/blob/master/docs/Support%20Docs/announcement-guidelines.md)).
- [ ] I prefer the discussion of this application to take place in a private Element/Matrix channel. My username is: `@_______:matrix.org` (change the homeserver if you use a different one)

### CODE REVIEW

**Comment by rbrandao** [comment]
I have read and hereby sign the Contributor License Agreement.

**Comment by rbrandao** [comment]
recheck

File: applications/polkadot_analytics_platform_stage2.md [hunk 1]
```diff
@@ -0,0 +1,420 @@
...
+This proposal is aligned to the [RFP: Data Analysis Tools for Substrate-based Blockchains](https://github.com/w3f/Grants-Program/blob/master/docs/RFPs/Under%20Development/data_analysis_tools.md). 
+
+Our R&D roadmap was conceived and kick-started before the creation of this RFP, which was presented to us during the course of our previous [research grant](https://github.com/w3f/Grants-Program/pull/1420). Nevertheless, we believe the proposed application (Polkadot Analytics Platform) is aligned with the RFP's requirements and main goals. Both aim at providing data analysis in a way that advanced non-technical users can consume, without requiring programming skills. 
+
+This is a follow-up grant application for the project [A Knowledge-Oriented Approach to Enhance Integration and Communicability in the Polkadot Ecosystem](https://github.com/w3f/Grants-Program/pull/1420) and the project [A Polkadot Analytics Platform: Stage 1](https://github.com/w3f/Grants-Program/pull/1883) 
+
+
+### Overview
+
+> tag line
```
Comment (nikw3f) [inline applications/polkadot_analytics_platform_stage2.md:21]:
```suggestion
**Tagline:** 
```

File: applications/polkadot_analytics_platform_stage2.md [hunk 2]
```diff
@@ -0,0 +1,420 @@
...
+
+Empowering the Polkadot community with a comprehensive analytics platform that provides natural language querying, supported by formal knowledge representations, and customizable dashboards for holistic data analyses.
+
+> A brief description of your project.
```
Comment (nikw3f) [inline applications/polkadot_analytics_platform_stage2.md:25]:
```suggestion
**A brief description of your project**
```

File: applications/polkadot_analytics_platform_stage2.md (hunk 1 above)
Comment (nikw3f) [reply applications/polkadot_analytics_platform_stage2.md:21]: `>` is used for code in markdown. Better use subtitle '##' or bold '**' syntax.

**Comment by rbrandao** [comment]
Thanks for the heads-up @nikw3f, I changed all the quotes (>) to bold (**) in the markdown format.

**Review by takahser — CHANGES_REQUESTED** [review]
@rbrandao I added a few inline comments, feel free to have a look.

File: applications/polkadot_analytics_platform_stage2.md [hunk 3]
```diff
@@ -0,0 +1,421 @@
...
+- **Total Costs:** 30,000 USD
+
+The total budget requested is US$ 30,000 for two PhDs dedicating about 30 days of work each during a 3 months project. Considering the average PhD **daily** rate of US$500 to design and develop a project.
+
+Considering 22 working days on average per month, in three months we have approximately 3 * 22 days = 66 workdays. Each PhD working 30 days gives a total of 60 worked days. The FTE then is 60 worked days / 66 workdays, which is approximately 0.91 FTE.
+
+
+### Milestone 1 — Data extraction mechanisms
+
+The main goal of this milestone is to ...
```
[the remark this answers is not in the document]
Comment (rbrandao) [reply applications/polkadot_analytics_platform_stage2.md]: Thanks for the comments @takahser. We adjusted the application file based on your feedback.

File: applications/polkadot_analytics_platform_stage2.md [hunk 4]
```diff
@@ -0,0 +1,421 @@
...
+The main goal of this milestone is to ... 
+
+- **Estimated Duration:** 1 month
+- **FTE:**  0.91
+- **Costs:** 10,000 USD
+
+| Number | Deliverable | Specification |
+| -----: | ----------- | ------------- |
+| **0a.** | Copyright and Licenses | Apache 2.0 |
+| **0b.** | Documentation/Tutorial | Detailed documentation of the data extraction mechanisms. |
```
[the remark this answers is not in the document]
Comment (rbrandao) [reply applications/polkadot_analytics_platform_stage2.md]: Detailed the documentation with python code docs and a jupyter nb explaining the use of the developed code.

File: applications/polkadot_analytics_platform_stage2.md [hunk 5]
```diff
@@ -0,0 +1,421 @@
...
+| **0c.** | Testing and Testing Guide | Tests and a guide on how to verify the correctness of data extraction mechanisms. |
+| **0d.** | Docker | Docker images for the data extraction mechanisms. |
+| **1** | Substrate-ETL extraction scripts | Extraction scripts to gather data from Substrate-ETL. |
+| **2** | Polkadot Data extraction scripts | Extraction scripts to gather polkadot related data not available on Substrate-ETL. |
+| **3** | GitHub Data extraction scripts | Extraction scripts to gather data from GitHub, which will be used to support queries relating to network engagement and innovation. |
```
[the remark this answers is not in the document]
Comment (rbrandao) [reply applications/polkadot_analytics_platform_stage2.md]: In the previous stage, we extended the POnto ontology to structure relevant concepts of the Polkadot ecosystem, including entities from example queries in the Substrate-ETL project and the related RFP mentioned in this application. 

In M1, the idea is to implement scripts to gather data to support the previously analyzed queries. That is, gathering data from Substrate-ETL, Substrate-based assets and services (such as Polkassembly, polkadot.js, and others), and GitHub (in order to answer questions about the ecosystem evolution).

In deliverable 1, we will extract data from all the SQL tables specified in the Substrate-ETL.
In deliverable 2, we will explore APIs like the ones used in the [Polkassembly](https://github.com/polkassembly/polkassembly/tree/main) to extract OpenGov data to answer the queries specified in the RFP. In deliverable 3, we will use github APIs to extract data from the repositories related to the Polkadot Ecosystem, aiming at answering questions regarding dev-community engagement. E.g.: which parachain had more PR in a specific time interval.

File: applications/polkadot_analytics_platform_stage2.md [hunk 6]
```diff
@@ -0,0 +1,421 @@
...
+The next stages in our roadmap, which are not in the scope of this application grant are:
+
+Stage 3: Controlled Natural Language (CNL) Engine
+In this stage, the roadmap foresees a controlled natural language (CNL) specification to leverage the POnto ontology. CNL is a simplified form of natural language designed to be easily understood by both humans and machines. The CNL specification will allow users to interact with the Polkadot Analytics Platform using natural language queries, making the platform more user-friendly and accessible.
+
+Stage 4: Query Engine
+This stage will focus on the development of a query engine to support the execution of CNL-based queries on the knowledge base. The query engine will be responsible for processing user queries, retrieving relevant data from the knowledge base, and providing users with query results.
+
+Stage 5: Visualization
+The final stage focuses on delivering a UI where users will be able to integrate multiple informative artifacts to compose their custom dashboards, facilitating personalized data exploration and analysis.
```
[the remark this answers is not in the document]
Comment (rbrandao) [reply applications/polkadot_analytics_platform_stage2.md:389]: Our current estimates for the next stages are:

Stage 3: Controlled Natural Language (CNL) Engine
- **Estimated Duration:** 2 months
- **FTE:**  0.91
- **Costs:** 20,000 USD

Stage 4: Query Engine
- **Estimated Duration:** 6 weeks
- **FTE:**  0.91
- **Costs:** 15,000 USD

Stage 5: Visualization - 2 Months
- **Estimated Duration:** 2 months
- **FTE:**  0.91
- **Costs:** 20,000 USD

[the remark this answers is not in the document]
Comment (rbrandao) [reply applications/polkadot_analytics_platform_stage2.md]: > How long do you estimate will these additional stages take and how much funding will you require to implement it? I'm just looking for ballpark figures here, to get a better understanding of the long-term roadmap.

I also added the funding estimates info to the "Future plans" section.

File: applications/polkadot_analytics_platform_stage2.md [hunk 7]
```diff
@@ -0,0 +1,440 @@
...
+
+| Number | Deliverable | Specification |
+| -----: | ----------- | ------------- |
+| **0a.** | Copyright and Licenses | Apache 2.0 |
+| **0b.** | Documentation/Tutorial | Documentation of the Python code using pdoc3. The tutorial will be a jupyter notebook explaining the use of the implemented semantic ETL workflows |
+| **0c.** | Testing and Testing Guide | Testing the information ETL process. A guide on how to verify the correctness of injected knowledge. |
+| **0d.** | Docker | Docker images for the ETL workflows. |
+| **0e.** | Article | We will write a medium post with an overview of semantic ETL workflows and their tasks. |
+| **1** | Concepts and data sources mapping | Extension of POnto, mapping relevant entities and corresponding datasources. This mapping will be used later on to automatize the proposed Semantic ETL Pipelines to continuously sync the KB with the latest ecosystem state. |
+| **2.** | Ontology Alignment | Scripts to perform ontology alignment with the extracted data as entities in the POnto ontology. |
```
Comment (semuelle) [inline applications/polkadot_analytics_platform_stage2.md:371]: What is ontology alignment?

**Review by semuelle — CHANGES_REQUESTED** [review]
Thanks for the application, @rbrandao. Could you add an architecture diagram that shows where and how the deliverables fit into the complete architecture? And ideally a deliverable in form of a demo showing how the results of this grant will be used in the future? Basically what the testing guide would be, but with the future work in mind ("This question will be converted in this query and this is the result of that query")?

**Comment by rbrandao** [comment]
> Thanks for the application, @rbrandao. Could you add an architecture diagram that shows where and how the deliverables fit into the complete architecture? 

Hi @semulle, we extended the architectural diagram (see Figure 2) reflecting your suggestions. The orange boxes highlight the components that we will develop in the current application. We are focusing on the Data Layer, to develop the extraction scripts (in M1) and to structure these scripts in "semantic workflows" (in M2) that will perform alignment with the POnto ontology and inject entities as individuals in the KB.

> And ideally a deliverable in form of a demo showing how the results of this grant will be used in the future? Basically what the testing guide would be, but with the future work in mind ("This question will be converted in this query and this is the result of that query")?

We updated the milestone 2 deliverable table, extending the tutorial (deliverable 0b) to comprise the demo you suggested in the Jupyter notebook. In addition, Figure 1 illustrates the main steps of the user interaction you mentioned (from the CNL query input to result visualization, and final user feedback).

File: applications/polkadot_analytics_platform_stage2.md (hunk 7 above)
Comment (rbrandao) [reply applications/polkadot_analytics_platform_stage2.md:374]: In our project, we consider "ontology alignment" the task of establishing correspondence between the POnto ontology and the schema structures used in different data sources. This process aims at enriching the extracted data with semantics, to allow structured queries and knowledge consumption. 

In general, ontologies are formal representations of knowledge that define the concepts and relationships within a specific domain. When extracting data from multiple sources, each source may employ its own ontology, leading to semantic heterogeneity. Ontology alignment aims to resolve these differences by mapping equivalent or related concepts across different sources, ensuring interoperability and coherence in the integrated data.

**Comment by rbrandao** [comment]
Hi @nikw3f, just saw that the github bot added the "stale" label to this PR. Anything pending on our side?

**Comment by takahser** [comment]
@rbrandao sorry for the delay here, we've currently got a bit of a backlog. I removed the label. We'll be back with more feedback soon!

**Comment by rbrandao** [comment]
> @rbrandao sorry for the delay here, we've currently got a bit of a backlog. I removed the label. We'll be back with more feedback soon!

Thanks for the prompt reply @takahser. We are looking forward for it.

**Review by semuelle — COMMENTED** [review]
Hi @rbrandao. Thanks for the updates. I just noticed that you are estimating another 55,000 USD required _after_ this grant to get to the final stage of the project. Including the previous grants, that's over 120,000 USD. Given the scope of other, much smaller grants from already established teams, this seems disproportionate. 
Since I won't be able to reply to comments in the next days, I will withhold my vote and let the rest of the committee give their view.

**Review by Noc2 — CHANGES_REQUESTED** [review]
Thanks a lot for the new application. However, after taking a look at your previous deliveries and your GitHub org, I decided not to support this application for now. In general, I think that UI is better suited for most of the aspects that the application tries to cover. Additionally, I wasn't convinced by your deliveries, nor do I see any progress since the delivery in your GitHub org. That said, I wish you all the best with the project and will ping the rest of the committee again.

**Comment by rbrandao** [comment]
> Hi @rbrandao. Thanks for the updates. I just noticed that you are estimating another 55,000 USD required _after_ this grant to get to the final stage of the project. Including the previous grants, that's over 120,000 USD. Given the scope of other, much smaller grants from already established teams, this seems disproportionate. Since I won't be able to reply to comments in the next days, I will withhold my vote and let the rest of the committee give their view.

Thanks for the feedback @semuelle. If I may ask, how much you think should be a proportional value for a similar scope of work that we are proposing? I mean, a broad solution encompassing mechanisms to structure diverse data sources, aligned through an ontology in a KB to be queried over a controlled natural language. And that led by a team of experts with proven record of delivering similar solutions in different industries. From our experience, 120k for such a project is below the market.

**Comment by rbrandao** [comment]
Hi @Noc2 , thanks for the comment.

> In general, I think that UI is better suited for most of the aspects that the application tries to cover. 

Maybe I'm missing something here? The UI (frontend) has to retrieve data from a source in the backend, as explained in our application we are pulling data from different sources (including chain data from ETLs) that would be aligned in a Knowledge Base. There is no way to get all of that on the UI without backend assets.

> Additionally, I wasn't convinced by your deliveries, nor do I see any progress since the delivery in your GitHub org. 

That is really unfortunate that you aren't convinced by our deliveries. We delivered all of the proposed milestones according to the roadmap defined in our previous applications, and we did not miss any deadline whatsoever. Regarding the progress on our github org, we have been working on private projects in our startup. I don´t think it is fair to assess the progress of a project that wasn't approved in the first place.

**Comment by takahser** [comment]
@rbrandao thanks for your patience and the work you put into this. However, unfortunately the w3f grants committee decided not to pursue your proposal further. The reasons for this decision include:

- the amount of funds you'd require to finish your projects (>$120k) in comparison to the value that we perceive is being created for the community in return
- the vague formulation of some parts of the proposal
- concerns related to the deliveries in your previous grant

At this point, getting your project funded through alternative funding sources like the treasury seem worth exploring. We'd like to add that you're still welcome to apply for grants at our program in the future.

We wish you all the best for the future of your project and thank you for your interest in our program!

**Pull request closed without merging**